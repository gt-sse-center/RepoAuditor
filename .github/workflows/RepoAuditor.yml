name: RepoAuditor

on: 
  workflow_dispatch:
  schedule:
    - cron: '0 0 * * *' # Every day at midnight UTC

jobs:
  audit-repo:
    runs-on: ubuntu-latest
    steps:
      - uses: mtanneau/RepoAuditor@action-v0.2.1
        with:
          github_pat: ${{ secrets.REPO_AUDITOR_PAT }}
