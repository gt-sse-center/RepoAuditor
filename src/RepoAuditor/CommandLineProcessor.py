# -------------------------------------------------------------------------------
# |
# |  Copyright (c) 2024 Scientific Software Engineering Center at Georgia Tech
# |  Distributed under the MIT License.
# |
# -------------------------------------------------------------------------------
"""Contains the CommandLineProcessor object"""

from dataclasses import dataclass, field
from typing import Any, Protocol

from dbrownell_Common.Streams.DoneManager import DoneManager  # type: ignore[import-untyped]

from .ExecuteModules import Execute, Module, ModuleInfo


# ----------------------------------------------------------------------
@dataclass(frozen=True)
class CommandLineProcessor:
    """\
    Object that makes it easy to process command line arguments and invoke ExecuteModules.Execute
    in a testable way.
    """

    # ----------------------------------------------------------------------
    # |
    # |  Public Types
    # |
    # ----------------------------------------------------------------------
    class GetDynamicArgsFunc(Protocol):
        def __call__(
            self,
            dynamic_arg_definitions: dict[
                str, Any
            ],  # actual type is dict[str, TyperEx.TypeDefinitionItemType]
        ) -> dict[str, Any]: ...

    # ----------------------------------------------------------------------
    # |
    # |  Public Data
    # |
    # ----------------------------------------------------------------------
    module_infos: list[ModuleInfo]
    warnings_as_error_module_names: set[str]
    ignore_warnings_module_names: set[str]
    single_threaded: bool = field(kw_only=True)

    # ----------------------------------------------------------------------
    # |
    # |  Public Methods
    # |
    # ----------------------------------------------------------------------
    @classmethod
    def Create(
        cls,
        get_dynamic_args_func: "CommandLineProcessor.GetDynamicArgsFunc",
        modules: list[Module],
        includes: list[str],
        excludes: list[str],
        warnings_as_error_module_names: set[str],
        ignore_warnings_module_names: set[str],
        *,
        all_warnings_as_error: bool = False,
        ignore_all_warnings: bool = False,
        single_threaded: bool = False,
        argument_separator: str = "-",
    ) -> "CommandLineProcessor":
        # Convert modules into a lookup map
        module_map: dict[str, Module] = {}

        for module in modules:
            if argument_separator in module.name:
                raise Exception(f"The module name '{module.name}' contains '{argument_separator}'.")

            prev_module = module_map.get(module.name, None)
            if prev_module is not None:
                raise Exception(f"The module '{module.name}' has already been defined.")

            module_map[module.name] = module

        del modules

        # Process includes
        included_modules: set[str] = set()
        included_requirements: dict[str, set[str]] = {}

        for include in includes:
            parts = include.split(argument_separator)

            this_module = module_map.get(parts[0], None)
            if this_module is None:
                raise Exception(f"'{parts[0]}' is not a recognized module name.")

            if len(parts) == 1:
                included_modules.add(parts[0])
            else:
                included_requirements.setdefault(this_module.name, set()).add(
                    argument_separator.join(parts[1:])
                )

        del includes

        for module_name, module in list(module_map.items()):
            if module.requires_explicit_include and module_name not in included_modules:
                module_map.pop(module_name)
                continue

        del included_modules

        # Process excludes
        excluded_requirements: dict[str, set[str]] = {}

        for exclude in excludes:
            parts = exclude.split(argument_separator)

            this_module = module_map.get(parts[0], None)
            if this_module is None:
                raise Exception(f"'{parts[0]}' is not a recognized module name.")

            if len(parts) == 1:
                module_map.pop(parts[0])
            else:
                excluded_requirements.setdefault(this_module.name, set()).add(
                    argument_separator.join(parts[1:])
                )

        del excludes

        # Add / Remove the requirements for each module
        for module_name, module in list(module_map.items()):
            module.ProcessRequirements(
                included_requirements.get(module_name, set()),
                excluded_requirements.get(module_name, set()),
            )

            if module.GetNumRequirements() == 0:
                module_map.pop(module_name)

        del included_requirements
        del excluded_requirements

        if all_warnings_as_error:
            warnings_as_error_module_names = {module.name for module in module_map.values()}
        if ignore_all_warnings:
            ignore_warnings_module_names = {module.name for module in module_map.values()}

        # Create the type definitions
        dynamic_arg_definitions: dict[str, Any] = {}

        for module in module_map.values():
            # Module-level args
            for key, value in module.GetDynamicArgDefinitions().items():
                dynamic_arg_definitions[f"{module.name}{argument_separator}{key}"] = value

            # Requirement-level args
            for query in module.queries:
                for requirement in query.requirements:
                    for key, value in requirement.GetDynamicArgDefinitions().items():
                        dynamic_arg_definitions[
                            f"{module.name}{argument_separator}{requirement.name}{argument_separator}{key}"
                        ] = value

        # Process the dynamic info
        dynamic_args: dict[str, dict[str, Any]] = {}

        for key, value in get_dynamic_args_func(dynamic_arg_definitions).items():
            parts = key.split(argument_separator)
            assert len(parts) >= 2

            if parts[0] not in module_map:
                raise Exception(f"'{parts[0]}' is not a recognized module name.")

            if len(parts) == 2:
                dynamic_args.setdefault(parts[0], {})[parts[1]] = value
            else:
                dynamic_args.setdefault(parts[0], {}).setdefault(None, {}).setdefault(  # type: ignore
                    parts[1],
                    {},
                )[argument_separator.join(parts[2:])] = value

        module_infos: list[ModuleInfo] = []

        for module_name, module in module_map.items():
            module_args = dynamic_args.get(module_name, {})
            requirement_args = module_args.pop(None, None)  # type: ignore

            module_infos.append(ModuleInfo(module, module_args, requirement_args))

        return cls(
            module_infos,
            warnings_as_error_module_names,
            ignore_warnings_module_names,
            single_threaded=single_threaded,
        )

    # ----------------------------------------------------------------------
    def __call__(
        self,
        dm: DoneManager,
    ) -> list[list[Module.EvaluateInfo]]:
        return Execute(dm, self.module_infos)
