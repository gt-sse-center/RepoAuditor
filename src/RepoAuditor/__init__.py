# pylint: disable=missing-module-docstring,invalid-name

APP_NAME = "RepoAuditor"

# Note that this value will be overwritten by calls to `python ../../Build.py update_version` based
# on changes observed in the git repository. The default value below will be used until the value
# here is explicitly updated by the Continuous Integration system.
__version__ = "0.1.24"
