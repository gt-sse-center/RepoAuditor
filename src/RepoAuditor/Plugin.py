# -------------------------------------------------------------------------------
# |
# |  Copyright (c) 2024 Scientific Software Engineering Center at Georgia Tech
# |  Distributed under the MIT License.
# |
# -------------------------------------------------------------------------------
"""Contains the interface that Plugins must implement."""

import pluggy

from RepoAuditor import APP_NAME
from RepoAuditor.Module import Module


# ----------------------------------------------------------------------
@pluggy.HookspecMarker(APP_NAME)
def GetModule() -> Module:
    """Return a Module."""
    raise NotImplementedError("hookspec")  # pragma: no cover # noqa: EM101
