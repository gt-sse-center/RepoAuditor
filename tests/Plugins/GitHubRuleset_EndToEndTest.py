# -------------------------------------------------------------------------------
# |
# |  Copyright (c) 2024 Scientific Software Engineering Center at Georgia Tech
# |  Distributed under the MIT License.
# |
# -------------------------------------------------------------------------------
"""End-to-end tests for the GitHubRulesets plugin."""

from pathlib import Path
import pytest

from dbrownell_Common.TestHelpers.StreamTestHelpers import (
    InitializeStreamCapabilities,
)
from typer.testing import CliRunner

from RepoAuditor.EntryPoint import app

from utilities import ScrubDurationGithuburlAndSpaces, GetGithubUrl, CheckPATFileExists

# ----------------------------------------------------------------------
pytest.fixture(InitializeStreamCapabilities(), scope="session", autouse=True)


# ----------------------------------------------------------------------
class TestGitHubRuleset:
    """End-to-end tests for rulesets in a GitHub repository."""

    def test_RequireStatusChecks(self, pat_args, snapshot):
        """Test for Status Checks not required ruleset"""
        result = CliRunner().invoke(app, pat_args + ["--GitHubRulesets-RequireStatusChecks-true"])

        assert result.exit_code == -1, result.output
        assert ScrubDurationGithuburlAndSpaces(result.stdout) == snapshot

    def test_RequirePullRequests(self, pat_args, snapshot):
        """Test for Pull Requests required ruleset"""
        result = CliRunner().invoke(app, pat_args + ["--GitHubRulesets-RequirePullRequests-true"])

        assert result.exit_code == -1, result.output
        assert ScrubDurationGithuburlAndSpaces(result.stdout) == snapshot

    def test_RequireSignedCommits(self, pat_args, snapshot):
        """Test for Signed Commits required ruleset"""
        result = CliRunner().invoke(app, pat_args + ["--GitHubRulesets-RequireSignedCommits-true"])

        assert result.exit_code == -1, result.output
        assert ScrubDurationGithuburlAndSpaces(result.stdout) == snapshot


# ----------------------------------------------------------------------
# ----------------------------------------------------------------------
# ----------------------------------------------------------------------
@pytest.fixture
def args() -> list[str]:
    return [
        "--include",
        "GitHubRulesets",
        "--GitHubRulesets-url",
        GetGithubUrl(),
    ]


# ----------------------------------------------------------------------
@pytest.fixture
def pat_args(args) -> list[str]:
    _github_pat_filename = (Path(__file__).parent / "github_pat.txt").resolve()
    CheckPATFileExists(_github_pat_filename)

    with _github_pat_filename.open() as f:
        pat_value = f.read().strip()

    return args + ["--GitHubRulesets-pat", pat_value]
