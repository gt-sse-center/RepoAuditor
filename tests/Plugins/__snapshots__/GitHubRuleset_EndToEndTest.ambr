# serializer version: 1
# name: TestGitHubRuleset.test_RequirePullRequests
  '''
  Processing 1 module...
    Processing 'GitHubRulesets' (1 of 1)...
  
  
  
  
    DONE! ([31;1m-1[0m, <scrubbed duration>)
  DONE! ([31;1m-1[0m, <scrubbed duration>)
  
  
  â•­â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€ GitHubRulesets â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•®
  â”‚                                                                                                                                                                                  â”‚
  â”‚ Validates GitHub repository rulesets.                                                                                                                                            â”‚
  â”‚                                                                                                                                                                                  â”‚
  â”‚ Successful:     0 (0.00%)                                                                                                                                                        â”‚
  â”‚ Warnings:       0 (0.00%)                                                                                                                                                        â”‚
  â”‚ Errors:         1 (33.33%)                                                                                                                                                       â”‚
  â”‚ Skipped:        2 (66.67%)                                                                                                                                                       â”‚
  â”‚                                                                                                                                                                                  â”‚
  â”‚ [31mâ•­â”€[0m[31m [Error] RequirePullRequests [0m[31mâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€[0m[31mâ”€â•®[0m â”‚
  â”‚ [31mâ”‚[0m                                                                                                                                                                              [31mâ”‚[0m â”‚
  â”‚ [31mâ”‚[0m Validates that 'pull_request' is set to the expected value.                                                                                                                  [31mâ”‚[0m â”‚
  â”‚ [31mâ”‚[0m                                                                                                                                                                              [31mâ”‚[0m â”‚
  â”‚ [31mâ”‚[0m ERROR: No active branch ruleset requiring Pull Requests found                                                                                                                [31mâ”‚[0m â”‚
  â”‚ [31mâ”‚[0m                                                                                                                                                                              [31mâ”‚[0m â”‚
  â”‚ [31mâ”‚[0m Resolution                                                                                                                                                                   [31mâ”‚[0m â”‚
  â”‚ [31mâ”‚[0m ----------                                                                                                                                                                   [31mâ”‚[0m â”‚
  â”‚ [31mâ”‚[0m   Enable pull request requirements in repository rulesets                                                                                                                    [31mâ”‚[0m â”‚
  â”‚ [31mâ”‚[0m                                                                                                                                                                              [31mâ”‚[0m â”‚
  â”‚ [31mâ”‚[0m Rationale                                                                                                                                                                    [31mâ”‚[0m â”‚
  â”‚ [31mâ”‚[0m ---------                                                                                                                                                                    [31mâ”‚[0m â”‚
  â”‚ [31mâ”‚[0m   Pull request reviews help maintain code quality and collaboration                                                                                                          [31mâ”‚[0m â”‚
  â”‚ [31mâ”‚[0m                                                                                                                                                                              [31mâ”‚[0m â”‚
  â”‚ [31mâ•°â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•¯[0m â”‚
  â”‚                                                                                                                                                                                  â”‚
  â•°â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•¯
  
  Results: DONE! ([31;1m-1[0m, <scrubbed duration>)
  
  
  '''
# ---
# name: TestGitHubRuleset.test_RequireSignedCommits
  '''
  Processing 1 module...
    Processing 'GitHubRulesets' (1 of 1)...
  
  
  
  
    DONE! ([31;1m-1[0m, <scrubbed duration>)
  DONE! ([31;1m-1[0m, <scrubbed duration>)
  
  
  â•­â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€ GitHubRulesets â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•®
  â”‚                                                                                                                                                                                  â”‚
  â”‚ Validates GitHub repository rulesets.                                                                                                                                            â”‚
  â”‚                                                                                                                                                                                  â”‚
  â”‚ Successful:     0 (0.00%)                                                                                                                                                        â”‚
  â”‚ Warnings:       0 (0.00%)                                                                                                                                                        â”‚
  â”‚ Errors:         1 (33.33%)                                                                                                                                                       â”‚
  â”‚ Skipped:        2 (66.67%)                                                                                                                                                       â”‚
  â”‚                                                                                                                                                                                  â”‚
  â”‚ [31mâ•­â”€[0m[31m [Error] RequireSignedCommits [0m[31mâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€[0m[31mâ”€â•®[0m â”‚
  â”‚ [31mâ”‚[0m                                                                                                                                                                              [31mâ”‚[0m â”‚
  â”‚ [31mâ”‚[0m Validates that 'required_signatures' is set to the expected value.                                                                                                           [31mâ”‚[0m â”‚
  â”‚ [31mâ”‚[0m                                                                                                                                                                              [31mâ”‚[0m â”‚
  â”‚ [31mâ”‚[0m ERROR: No active branch ruleset requiring Signed Commits found                                                                                                               [31mâ”‚[0m â”‚
  â”‚ [31mâ”‚[0m                                                                                                                                                                              [31mâ”‚[0m â”‚
  â”‚ [31mâ”‚[0m Resolution                                                                                                                                                                   [31mâ”‚[0m â”‚
  â”‚ [31mâ”‚[0m ----------                                                                                                                                                                   [31mâ”‚[0m â”‚
  â”‚ [31mâ”‚[0m   Enable commit signing requirement in repository rulesets                                                                                                                   [31mâ”‚[0m â”‚
  â”‚ [31mâ”‚[0m                                                                                                                                                                              [31mâ”‚[0m â”‚
  â”‚ [31mâ”‚[0m Rationale                                                                                                                                                                    [31mâ”‚[0m â”‚
  â”‚ [31mâ”‚[0m ---------                                                                                                                                                                    [31mâ”‚[0m â”‚
  â”‚ [31mâ”‚[0m   Signed commits ensure commit authenticity                                                                                                                                  [31mâ”‚[0m â”‚
  â”‚ [31mâ”‚[0m                                                                                                                                                                              [31mâ”‚[0m â”‚
  â”‚ [31mâ•°â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•¯[0m â”‚
  â”‚                                                                                                                                                                                  â”‚
  â•°â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•¯
  
  Results: DONE! ([31;1m-1[0m, <scrubbed duration>)
  
  
  '''
# ---
# name: TestGitHubRuleset.test_RequireStatusChecks
  '''
  Processing 1 module...
    Processing 'GitHubRulesets' (1 of 1)...
  
  
  
  
    DONE! ([31;1m-1[0m, <scrubbed duration>)
  DONE! ([31;1m-1[0m, <scrubbed duration>)
  
  
  â•­â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€ GitHubRulesets â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•®
  â”‚                                                                                                                                                                                  â”‚
  â”‚ Validates GitHub repository rulesets.                                                                                                                                            â”‚
  â”‚                                                                                                                                                                                  â”‚
  â”‚ Successful:     0 (0.00%)                                                                                                                                                        â”‚
  â”‚ Warnings:       0 (0.00%)                                                                                                                                                        â”‚
  â”‚ Errors:         1 (33.33%)                                                                                                                                                       â”‚
  â”‚ Skipped:        2 (66.67%)                                                                                                                                                       â”‚
  â”‚                                                                                                                                                                                  â”‚
  â”‚ [31mâ•­â”€[0m[31m [Error] RequireStatusChecks [0m[31mâ”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€[0m[31mâ”€â•®[0m â”‚
  â”‚ [31mâ”‚[0m                                                                                                                                                                              [31mâ”‚[0m â”‚
  â”‚ [31mâ”‚[0m Validates that 'required_status_checks' is set to the expected value.                                                                                                        [31mâ”‚[0m â”‚
  â”‚ [31mâ”‚[0m                                                                                                                                                                              [31mâ”‚[0m â”‚
  â”‚ [31mâ”‚[0m ERROR: No active branch ruleset requiring Status Checks found                                                                                                                [31mâ”‚[0m â”‚
  â”‚ [31mâ”‚[0m                                                                                                                                                                              [31mâ”‚[0m â”‚
  â”‚ [31mâ”‚[0m Resolution                                                                                                                                                                   [31mâ”‚[0m â”‚
  â”‚ [31mâ”‚[0m ----------                                                                                                                                                                   [31mâ”‚[0m â”‚
  â”‚ [31mâ”‚[0m   Configure required status checks in repository rulesets                                                                                                                    [31mâ”‚[0m â”‚
  â”‚ [31mâ”‚[0m                                                                                                                                                                              [31mâ”‚[0m â”‚
  â”‚ [31mâ”‚[0m Rationale                                                                                                                                                                    [31mâ”‚[0m â”‚
  â”‚ [31mâ”‚[0m ---------                                                                                                                                                                    [31mâ”‚[0m â”‚
  â”‚ [31mâ”‚[0m   Status checks ensure code quality and compatibility                                                                                                                        [31mâ”‚[0m â”‚
  â”‚ [31mâ”‚[0m                                                                                                                                                                              [31mâ”‚[0m â”‚
  â”‚ [31mâ•°â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•¯[0m â”‚
  â”‚                                                                                                                                                                                  â”‚
  â•°â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•¯
  
  Results: DONE! ([31;1m-1[0m, <scrubbed duration>)
  
  
  '''
# ---
